### PR TITLE
Add changes to compile rust-gpu on the latest nightly

### PR DESCRIPTION
**Pull request by XAMPPRocky**
This PR along with allows rustc to be able to pull in and compile rust-gpu as a codegen backend. There's also some changes that are needed on rustc side, for which I'll create a follow up PR that contains a way to checkout and patch rust-lang/rust.

### CODE REVIEW

**Comment by MarijnS95** [comment]
@XAMPPRocky The next sources changed `Scalar` handling which I've patched to see if the latest rustc would resolve another issue, perhaps this is useful to you: https://github.com/EmbarkStudios/rust-gpu/compare/pub-changes...MarijnS95:nightly

**Comment by bjorn3** [comment]
If you symlink it to `compiler/rustc_codegen_spirv`, it should be possible to build it using `codegen-backends = ["llvm", "spirv"]`. The bootstrap changes to compile cg_clif should work with any codegen backend that can be found at `compiler/rustc_codegen_$name`.

File: crates/rustc_codegen_spirv/src/lib.rs [hunk 1]
```diff
@@ -206,7 +206,7 @@ impl MetadataLoader for SpirvMetadataLoader {
 }
 
 #[derive(Clone)]
-struct SpirvCodegenBackend;
+pub struct SpirvCodegenBackend;
```
Comment (khyperia) [inline crates/rustc_codegen_spirv/src/lib.rs:209]: Why are these pub now?

**Review by khyperia — COMMENTED** [review]
Still need to wait for `rustfmt` to build on a new nightly and whatnot, but looks good pending that.

**Comment by XAMPPRocky** [comment]
@bjorn3 Thanks for that, I tried using that method, however when trying that it fails on building the compiler_artifacts with a seemingly unrelated error.

```
error[E0107]: wrong number of type arguments: expected 2, found 1
   --> /Users/src/rust/compiler/rustc_data_structures/src/graph/iterate/mod.rs:271:10
    |
271 |     ) -> ControlFlow<Self::BreakVal> {
    |          ^^^^^^^^^^^^^^^^^^^^^^^^^^^ expected 2 type arguments

error[E0107]: wrong number of type arguments: expected 2, found 1
   --> /Users/src/rust/compiler/rustc_data_structures/src/graph/iterate/mod.rs:276:51
    |
276 |     fn node_settled(&mut self, _node: G::Node) -> ControlFlow<Self::BreakVal> {
    |                                                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^ expected 2 type arguments

error[E0107]: wrong number of type arguments: expected 2, found 1
   --> /Users/src/rust/compiler/rustc_data_structures/src/graph/iterate/mod.rs:299:10
    |
299 |     ) -> ControlFlow<Self::BreakVal> {
    |          ^^^^^^^^^^^^^^^^^^^^^^^^^^^ expected 2 type arguments
```

**Comment by bjorn3** [comment]
https://github.com/rust-lang/rust/blob/f58ffc93815f76576eb56df4bdeec2fe8f12b766/library/core/src/ops/control_flow.rs#L6

Huh, the second type argument has a default. Can you post the full build log? Maybe that helps?

**Comment by XAMPPRocky** [comment]
[Build Log](https://paste.rs/OEc)

**Comment by bjorn3** [comment]
> command did not execute successfully: "/Users/src/rust/build/x86_64-apple-darwin/stage0/bin/cargo" "build" "--target" "x86_64-apple-darwin" "-Zbinary-dep-depinfo" "-j" "16" "--release" "--manifest-path" "/Users/src/rust/compiler/**rustc_codegen_spirv**/Cargo.toml" "--message-format" "json-render-diagnostics"

Did you add a dependency on `rustc_data_structures` or other rustc crates to `Cargo.toml` of `rustc_codegen_spirv`? That is not necessary and will cause issues.

**Comment by XAMPPRocky** [comment]
@bjorn3 I added it because not having it gives the following error.

```
   Compiling spirv-builder v0.1.0 (/Users/src/rust/compiler/rustc_codegen_spirv/crates/spirv-builder)
error[E0463]: can't find crate for `rustc_ast`
  --> crates/rustc_codegen_spirv/src/lib.rs:57:1
   |
57 | extern crate rustc_ast;
   | ^^^^^^^^^^^^^^^^^^^^^^^ can't find crate

error: aborting due to previous error

For more information about this error, try `rustc --explain E0463`.
error: could not compile `rustc_codegen_spirv`
```

https://paste.rs/8DF

**Comment by bjorn3** [comment]
Only `crates/rustc_codegen_spirv` needs to be symlinked I think. (likely requires a `Cargo.toml` change to fix the `spirv-builder` path I think) Even the way you do it right now that error shouldn't happen I think.

**Comment by XAMPPRocky** [comment]
@bjorn3 Well we do actually need crates as `rustc_codegen_spirv` depends on `spire-tools`, however symlinking just `rustc_codegen_spirv` and changing the path to the dependency manually in spirv's Cargo.toml does cause it to start working, I'm running through compiletest right now and will report more when it finishes.

**Comment by MarijnS95** [comment]
Good news, 2020-11-09 [includes `rustfmt` again](https://rust-lang.github.io/rustup-components-history/) :tada: 

Fixes for 2020-11-06 and 2020-11-09 pushed to https://github.com/EmbarkStudios/rust-gpu/compare/pub-changes...MarijnS95:nightly

**Comment by XAMPPRocky** [comment]
@MarijnS95 Nice, I ended up implementing the changes before I saw your comment. 😄

**Comment by MarijnS95** [comment]
@XAMPPRocky I figured as much, saw your rebase and had to act "quick" as I had already done these changes over the weekend, should have mentioned that before :grimacing: 

At least we both sorted the members in `TargetOptions` :tada:

**Comment by XAMPPRocky** [comment]
> EDIT: Btw these properties were moved from Target to TargetOptions because the latter one has sensible defaults. Most of our target_ fields match up, do you think we should omit some?

I've updated it now so that any that matched were removed. The only one remaining I believe is `target_os` since I don't know if we want the target name to be `spirv-unknown-none` or `spirv-unknown-unknown`, the former feels more accurate but I don't it matters much.

**Comment by MarijnS95** [comment]
> I've updated it now so that any that matched were removed. The only one remaining I believe is `target_os` 

Yeah I came to the same conclusion, looks good this way :ok_hand: 

> since I don't know if we want the target name to be `spirv-unknown-none` or `spirv-unknown-unknown`, the former feels more accurate but I don't think it matters much.

Hmm, that's a tough one. Only `wasm32` seems to be using `unknown` for the OS, and searching around Clang appears to use `spir64-unknown-unknown`?

**Comment by bjorn3** [comment]
You could say that the OS spirv code is running on is the driver or firmware, so in that case `unknown` makes sense, as you aren't running it bare bones without any driver at all, nor is the spirv the firmware itself.

**Review by khyperia — APPROVED** [review]
:tada:

**Pull request merged**